### PR TITLE
Added Widgets

### PR DESCRIPTION
**Pull request by Nyoxis**
Added nested widgets code structure. Each widget should be independent and its coordinates encapsulated, though global render target used. Added working fast and partial display refresh methods and sequential rendering. New pin entering display using new features.

### CODE REVIEW

File: kampela-system/src/devices/display copy2.rs [hunk 1]
```diff
@@ -0,0 +1,986 @@
+//! display control functions
```
Comment (Slesarew) [inline kampela-system/src/devices/display copy2.rs:1]: can we just delete this now?

File: kampela-ui/src/main.rs [hunk 2]
```diff
@@ -85,7 +93,7 @@ impl HALHandle {
 
 #[derive(Debug)]
 struct DesktopSimulator {
-    pin: Pincode,
+    pin: [u8; 4],
```
Comment (Slesarew) [inline kampela-ui/src/main.rs:96]: why not hard type/object?

File: kampela-ui/src/pin/pin.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    pindots: Pindots,
+    entered_nums: Vec<u8>,
+    pinok: bool,
+}
+
+impl<R> Pincode<R> where
+    R: Rng + ?Sized
+{
+    pub fn new(rng: &mut R) -> Self {
+        let pindots_size = Size {
```
Comment (Slesarew) [inline kampela-ui/src/pin/pin.rs:53]: is it better than leaving these in constants?

**Review by Slesarew — COMMENTED** [review]
Read the code; looks really good. Will test on device.

File: kampela-system/src/devices/display copy2.rs (hunk 1 above)
Comment (Nyoxis) [reply kampela-system/src/devices/display copy2.rs:1]: yes, this file was not deleted inadvertently

File: kampela-ui/src/main.rs (hunk 2 above)
Comment (Nyoxis) [reply kampela-ui/src/main.rs:96]: Probably. Better solution for pin code type, storage and initialization needs consideration

File: kampela-ui/src/pin/pin.rs (hunk 3 above)
Comment (Nyoxis) [reply kampela-ui/src/pin/pin.rs:53]: Better to use precomputed constants

**Pull request closed without merging**